### PR TITLE
[v0.8] feat(externAuthz): bring external authorization

### PR DESCRIPTION
**Pull request by eduser25**
This commit bring configurable inbound external authorization
to OSM v0.8.

Signed-off-by: Eduard Serra <eduser25@gmail.com>

- New Functionality      [x]

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
No

### CODE REVIEW

File: charts/osm/values.yaml [hunk 1]
```diff
@@ -131,6 +131,15 @@ OpenServiceMesh:
     # -- Destination's API or collector endpoint where the spans will be sent to
     endpoint: "/api/v2/spans"
 
+  # Inbound external authorization
```
Comment (shashankram) [inline charts/osm/values.yaml:134]: This comment should probably be a bit more verbose.

File: docs/content/docs/external_auth_opa.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  inbound_extauthz_enable: true
+  inbound_extauthz_address: opa.opa.svc.cluster.local
+  inbound_extauthz_port: 9191
+  inbound_extauthz_statprefix: authz_opa
+  inbound_extauthz_timeout: 1s
+  inbound_extauthz_failuremodeallow: false
+```
+
+- Uses (as `opa-envoy-plugin`) a configmap to set OPA's policy. [This is not intended for production](https://github.com/open-policy-agent/opa-envoy-plugin#example-bundle-configuration). 
+- `opa-envoy-plugin` does not seem to react to changes on the configmap (is this intended?). To update policies, currently the OPA container needs to be restarted _(I believe I'm missing something here)_.
```
Comment (shashankram) [inline docs/content/docs/external_auth_opa.md:30]: Clarify ambiguity with a github issue?

File: docs/content/docs/external_auth_opa.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+```
+kubectl edit configmap opa-policy -n opa
+```
+change specifically the default-all from:
+```
+default allow = false
+```
+to
+```
+default allow = true
```
Comment (shashankram) [inline docs/content/docs/external_auth_opa.md:85]: This seems to be updating the default. Is there an ask to support more granular policies?

File: pkg/envoy/lds/connection_manager.go [hunk 4]
```diff
@@ -15,11 +15,20 @@ import (
...
+// perspective, in which the connection manager filters will be applied
+type ConnManagerFilterDirection string
+
 const (
 	statPrefix = "http"
+
+	// Incoming defines in-mesh inbound and ingress traffic driections
+	Incoming = "incoming"
+	// Outgoing defines in-mesh outbound and egress traffic directions
+	Outgoing = "outgoing"
```
Comment (shashankram) [inline pkg/envoy/lds/connection_manager.go:28]: nit: use `inbound, outbound` to be consistent with other parts of the code (ex. RDS route direction).

Also un-export the const.

File: pkg/envoy/lds/connection_manager.go [hunk 5]
```diff
@@ -15,11 +15,20 @@ import (
 	"github.com/openservicemesh/osm/pkg/featureflags"
 )
 
+// ConnManagerFilterDirection defines, for filter terms, the direction of the traffic from an application
+// perspective, in which the connection manager filters will be applied
+type ConnManagerFilterDirection string
```
Comment (shashankram) [inline pkg/envoy/lds/connection_manager.go:20]: Just `connectionDirection` perhaps?
Note: un-export the type.

File: pkg/envoy/lds/ingress.go [hunk 6]
```diff
@@ -38,7 +38,7 @@ func newIngressHTTPFilterChain(cfg configurator.Configurator, svc service.MeshSe
 		return nil
 	}
 
-	inboundConnManager := getHTTPConnectionManager(route.InboundRouteConfigName, cfg, nil)
+	inboundConnManager := getHTTPConnectionManager(route.InboundRouteConfigName, cfg, nil, Incoming)
```
Comment (shashankram) [inline pkg/envoy/lds/ingress.go:41]: What's the role of external auth in Ingress? Is this required for ingress?

File: docs/content/docs/external_auth_opa.md (hunk 3 above)
Comment (eduser25) [reply docs/content/docs/external_auth_opa.md:85]: there is, but it's a matter of writting the polici down in Rego format. For the sake of simplicity I assumed just tweaking a default allow/do-not-allow var

Comment (eduser25) [reply docs/content/docs/external_auth_opa.md:85]: ref: https://www.openpolicyagent.org/docs/latest/policy-language/

File: pkg/envoy/lds/connection_manager.go (hunk 4 above)
Comment (eduser25) [reply pkg/envoy/lds/connection_manager.go:28]: I'll talk to you offline, had a Q about this.

File: charts/osm/values.yaml (hunk 1 above)
Comment (eduser25) [reply charts/osm/values.yaml:134]: will do

File: docs/content/docs/external_auth_opa.md (hunk 2 above)
Comment (eduser25) [reply docs/content/docs/external_auth_opa.md:30]: oh.... I need to clean up this comments, stale.

File: docs/content/docs/external_auth_opa.md (hunk 3 above)
Comment (shashankram) [reply docs/content/docs/external_auth_opa.md:85]: Should we drop one more example for a more granular rule?

Comment (eduser25) [reply docs/content/docs/external_auth_opa.md:85]: I'll might augment that on `main`, but I plan to leave it like this in this branch for now

**Review by snehachhabria — APPROVED** [review]

**Review by shashankram — APPROVED** [review]

**Pull request merged**